From 953558739672e294180a77baaf49285323244ce0 Mon Sep 17 00:00:00 2001
From: Max Horn <max@quendi.de>
Date: Thu, 13 Oct 2016 15:01:23 +0200
Subject: [PATCH] Fix compatibility with fplll 5.0.x

This also compiles with fplll 4.0.x, but not with older versions.
Resolves #43.
---
 linbox/algorithms/lattice.inl | 24 ++++++++++++------------
 1 file changed, 12 insertions(+), 12 deletions(-)

diff --git a/linbox/algorithms/lattice.inl b/linbox/algorithms/lattice.inl
index f29ed0b..e4ed890 100644
--- a/linbox/algorithms/lattice.inl
+++ b/linbox/algorithms/lattice.inl
@@ -324,6 +324,11 @@ namespace LinBox
 namespace LinBox
 {
 #ifdef __LINBOX_HAVE_FPLLL
+
+#if FPLLL_MAJOR_VERSION < 5
+#define lll_reduction lllReduction
+#endif
+
 	//! @bug we suppose Ring and mpz_t understand eachother...
 	template<class Ring, bool withU>
 	void
@@ -339,31 +344,26 @@ namespace LinBox
 		for (size_t i = 0 ; i < H.rowdim() ; ++i) {
 			for (size_t j = 0 ; j < H.coldim() ; ++j) {
                                 fplll::Z_NR<ZT> x = H.getEntry(i,j);
-                                B.Set((int)i,(int)j,x);
+                                B(i,j) = x;
 			}
 		}
 		// LLL()
 		switch (meth.getMeth()) {
 		case (latticeMethod::latticeFPLLL::P) :
 			{
-				fplll::proved<ZT,double> lllMethod(&B,meth.getPrecision(),
-							      meth.getEta(),meth.getDelta());
-				lllMethod.LLL();
+				lll_reduction(B, meth.getDelta(), meth.getEta(), fplll::LM_PROVED, fplll::FT_DOUBLE, meth.getPrecision());
 			}
 			break;
 		case (latticeMethod::latticeFPLLL::W) :
 			{
-				fplll::wrapper lllMethod(&B,meth.getPrecision(),
-						    meth.getEta(),meth.getDelta());
-				lllMethod.LLL();
+                                // precision is *ignored*
+				lll_reduction(B, meth.getDelta(), meth.getEta(), fplll::LM_WRAPPER);
 			}
 			break;
 		case (latticeMethod::latticeFPLLL::H) :
 			{
-				fplll::heuristic<ZT,double> lllMethod(&B,meth.getPrecision(),
-								 meth.getEta(),meth.getDelta(),
-								 meth.getSiegel());
-				lllMethod.LLL();
+                                int flags = meth.getSiegel() ? fplll::LLL_SIEGEL : fplll::LLL_DEFAULT;
+				lll_reduction(B, meth.getDelta(), meth.getEta(), fplll::LM_HEURISTIC, fplll::FT_DOUBLE, meth.getPrecision(), flags);
 			}
 			break;
 		}
@@ -373,7 +373,7 @@ namespace LinBox
 		// Convert back H
 		for (size_t i = 0 ; i < H.rowdim() ; ++i) {
 			for (size_t j = 0 ; j < H.coldim() ; ++j) {
-				H.setEntry(i,j, B.Get(i,j) );
+				H.setEntry(i,j, B(i,j) );
 			}
 		}
 
